### ISSUE TITLE
Proposal: New "isPartiallyHidden" Semantics Flag

### ISSUE THREAD

**Issue by nscobie** [flutter/flutter#62931]
## Links

Will fix issues for Android (https://github.com/flutter/flutter/issues/36307) and iOS (https://github.com/flutter/flutter/issues/61624) currently affecting `customer: money`.

Will require changes to the framework (https://github.com/flutter/flutter/pull/63031) and the engine (https://github.com/flutter/engine/pull/20285).

## Use case

When using an accessibility service, focusing on an element (by tap or swipe) which is partially offscreen in a scrollable container (`GridView`, `ListView`, etc.) should bring the focused element fully on screen.

### Native behavior on mobile platforms:

Tested with list and grid layout mode in Google Drive, Dropbox, and native "Files" apps for both platforms.

*   **Android:** Swiping right to a partially hidden element at the edge of a one-dimensional scrollable container (`ListView`) will trigger TalkBack to call `showOnScreen` for the element. For two-dimensional scrollable containers (`GridView`), this behavior is only seen when swiping to the _last cell_ of the last row. In both cases, tapping to focus does not trigger `showOnScreen`.
*   **iOS:** Swiping to a partially hidden element at the edge of any scrollable container (both `GridView` and `ListView`) **does** make the focused element visible by scrolling it to the center of the visible viewport of the container. Tapping to focus a partially hidden element does not bring it into view.

### Current Flutter behavior on mobile platforms:

*   **Android:** Appears to be the same as current Flutter behavior described above. I believe we should improve this to make `GridView` better.
*   **iOS:** Focusing on a partially hidden element at the edge of any scrollable container does **not** cause the element to be shown on screen.

### Proposed Flutter behavior on mobile platforms:

*   **Android:** Focusing a partially hidden element by swiping or tapping\* should trigger `showOnScreen`. _The current TalkBack limitation of not working for the final row of a `GridView` should be fixed._
*   **iOS:** Focusing a partially hidden element by swiping or tapping\* should trigger `showOnScreen`. _In the future, the call to `showOnScreen` should be easily replaceable by a call to some `scrollToMiddleOfView` function to fully mimic native iOS._

**\*Note:** see Q&A -> Open -> "Neither TalkBack nor VoiceOver will scroll to show a newly focused element if the element was focused by a tap, rather than a horizontal swipe. Should we mimic this, or "enhance" them to also scroll to elements focused via tap?"

## Proposal

Platforms should have an easy way to detect when a partially hidden element gains focus, and respond by generally mimicking native behavior for how the element should be brought on screen. However, there are some areas where native behavior can be improved, such as how TalkBack on Android handles GridViews.

This can be achieved by adding a new `isPartiallyHidden` [SemanticsFlag](https://api.flutter.dev/flutter/dart-ui/SemanticsFlag-class.html) in the framework to partially hidden nodes, allowing platforms to do a simple flag check when a SemanticsNode is focused on the engine side, and then respond appropriately (likely by calling `showOnScreen`).

If the engine were to lazily calculate whether a newly focused node is partially hidden, it would take `O(K)` time (where `K` = depth) every time a node is focused, and the operation would need to be re-implemented for every platform. By adding a flag to nodes in the framework we reduce re-implementation burden, and by extending an [existing process](https://github.com/flutter/flutter/blob/8081455394044dc4836c9c8fc11b5376dc3106bd/packages/flutter/lib/src/rendering/object.dart#L3808) we only add `O(1)` extra processing time per node during tree updates.

### Specific changes

This is an outline for specific changes to be made. See the PRs linked above for exact details.

#### Framework

<details>

The current [code which calculates](https://github.com/flutter/flutter/blob/8081455394044dc4836c9c8fc11b5376dc3106bd/packages/flutter/lib/src/rendering/object.dart#L3834) the [`isHidden` flag](https://api.flutter.dev/flutter/dart-ui/SemanticsFlag/isHidden-constant.html) should be extended to check if the `paintRect` intersection shrunk the previously calculated semantics rect, and set a temporary flag on the `SemanticsGeometry` of the node if so.

This temporary flag will later determine if the `isPartiallyHidden` flag should be added to the `SemanticsConfiguration` for a given node while the semantics tree is being compiled. This will mirror the [current code for setting the `isHidden` flag](https://github.com/flutter/flutter/blob/8081455394044dc4836c9c8fc11b5376dc3106bd/packages/flutter/lib/src/rendering/object.dart#L3682).

</details>

#### Engine - General

<details>

An entry for the new `isPartiallyHidden` semantics flag must be added to [semantics.dart](https://github.com/flutter/engine/blob/master/lib/ui/semantics.dart) and [semantics\_node.h](https://github.com/flutter/engine/blob/master/lib/ui/semantics/semantics_node.h), and the number of flags must be updated in [semantics\_test.dart](https://github.com/flutter/engine/blob/master/testing/dart/semantics_test.dart).

</details>

#### Engine - Android

<details>

When TalkBack calls [`performAction(...)` for `ACTION\_ACCESSIBILITY\_FOCUS` in AccessibilityBridge](https://github.com/flutter/engine/blob/3081844cb4dd4733f2162c7003021249e66e7036/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L937), a check for the new `isPartiallyHidden` flag on the semantics node should determine if the element should be shown on screen. The engine can then request that the framework bring it on screen by calling the `showOnScreen` semantics action for the node.

To avoid jank while scrolling with 2-3 fingers, we must also ensure the previously focused element does not have the `isHidden` flag.

Additionally, an entry for `isPartiallyHidden` must be added to the [`Flag` enum in AccessibilityBridge.java](https://github.com/flutter/engine/blob/3081844cb4dd4733f2162c7003021249e66e7036/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1702).

##### Known issues

*   Swiping to focus the boundary element in a `ListView will call `showOnScreen` twice (once by TalkBack, and then again by us). I don't believe this will be a problem, but could be checked for and avoided if it is.

</details>

#### Engine - iOS

<details>

When VoiceOver calls [`accessibilityElementDidBecomeFocused` on a `SemanticsObject`](https://github.com/flutter/engine/blob/0881bab52dcf32732db3ce42b8aeed5e0fde0acd/shell/platform/darwin/ios/framework/Source/SemanticsObject.mm#L453), the engine already checks if the node has the `isHidden` or `isHeader` semantic flags, and if so requests that the framework bring it on screen by calling the `showOnScreen` semantics action for the node. Another OR clause should be added to this condition to check if the focused element has the new `isPartiallyHidden` flag.

</details>

#### Engine - Other platforms (web, desktop, fuchsia, etc.)

<details>

I have not researched how these non-mobile platforms should handle responding to `isPartiallyHidden`. If this proposal is accepted, the implementation specifics for non-mobile platforms can be determined at a later date, with issues created to track progress.

</details>

#### Why not do this entirely in the engine, and avoid adding an additional flag?

<details>

- Whenever a new node gains focus, we could likely walk the tree back from the selected node up to the root, and at each step cumulatively apply the current node's transform to the original node's rect (accumulating the transformations), and check if it fits within the bounds of the parent's rect.

- However, this would require walking up the tree every time a node is selected, and this logic would need to be reimplemented for every platform.

</details>

#### Why not just do this entirely in the framework, so no additional code needs to be written for each platform?

<details>

- We could hook into the [onDidGainAccessibilityFocus](https://api.flutter.dev/flutter/semantics/SemanticsConfiguration/onDidGainAccessibilityFocus.html) callback for nodes which have the proposed `isPartiallyHidden` flag, and have the framework always call `showOnScreen` for them.

- However, different platforms will want to handle this differently. Android will need checks to ensure we don't jankily interrupt 2-3 finger scrolling, and iOS may eventually want to do something like scroll focused nodes to the middle of the screen (as seen with grid layout mode in the native Files app) instead of scrolling just enough to show the node on the edge of the screen.

- Further, future platforms may have native accessibility support for scrolling these elements onto screen. This proposal allows for platform specific control, while also improving code sharing.

</details>

#### Why can't we apply the `isHidden` flag to these partially hidden boundary nodes, rather than introducing a new flag?

<details>

- This would prevent iOS users from directly tapping partially hidden elements. More importantly, other platforms are encouraged to drop all `isHidden` semantics nodes from the semantics tree entirely.

</details>

#### Why can't we check if the next node has the `isHidden` flag to determine if we're a boundary node?

<details>

- This would only work correctly for 1D collections such as `ListView`. We don't know how many columns are present in 2D collections such as `GridView`, so we wouldn't know how far ahead in the list of nodes to look for one with `isHidden`.

</details>

#### Why name this flag `isPartiallyHidden`?

<details>

- What we're doing is similar to the `isHidden` flag, but only… _partially_.

- There is already an [`obscured` flag](https://api.flutter.dev/flutter/semantics/SemanticsProperties/obscured.html) which is used for password fields, so something like `isPartiallyObscured` would be confusing.

- Platforms may want to react differently to `isPartiallyHidden`, so something like `showOnFocus` or `scrollOnFocus` imply actions which would not always make sense.

- I'm very open to other names for this, feel free to make suggestions.

</details>

#### Is it okay to add another entry to the SemanticsFlag bitflag for something relatively small like this?

- Currently we are working with 32 bit integer flags ([semantics\_node.h](https://github.com/flutter/engine/blob/7a022774c285b2354520b6c679cc9161f4f9fe7a/lib/ui/semantics/semantics_node.h#L57)), is it possible to upgrade this to 64 if we run out?

#### The `isHidden` flag is exposed to end users via the [Semantics widget](https://api.flutter.dev/flutter/widgets/Semantics-class.html), should we do the same for `isPartiallyHidden`?

- Unsure of a use case for this, but it may prove helpful to some users.

#### How will this affect other widgets/semantic nodes which are not in scrollable containers but still partially obscured?

- Unsure if this situation would arise. If so, `showOnScreen` will hopefully be a no-op.

#### Neither TalkBack nor VoiceOver will scroll to show a newly focused element if the element was focused by a tap, rather than a horizontal swipe. Should we mimic this, or "enhance" them to also scroll to elements focused via tap?

- Perhaps users don't expect anything to move when tapping around, but they do when swiping? If so, this can be supported by:

    - **Android:** ensuring `hoveredObject == null`.

    - **iOS:** Unsure, TODO.

**Comment by goderbauer** [comment]
> When using an accessibility service, focusing on an element (by tap or swipe) which is partially offscreen in a scrollable container (GridView, ListView, etc.) should bring the focused element fully on screen.

No! This should only happen if you land there by swiping left/right. If the user explores the screen by sliding his finger over the screen (explore by touch), the content of the screen should not change as that would be very confusing. Therefore, if the focus lands on a partially visible node by sliding your finger over it it should not scroll the list.

Also, I believe when you just tap a node directly, it wouldn't scroll on native, so Flutter shouldn't either.

> Is it okay to add another entry to the SemanticsFlag bitflag for something relatively small like this?

It would be ok to add another flag, if we run out of space, we'll have to figure out a solution for how to make the space bigger.

> The isHidden flag is exposed to end users via the Semantics widget, should we do the same for isPartiallyHidden?

The Semantics widget should expose all flags.

> Neither TalkBack nor VoiceOver will scroll to show a newly focused element if the element was focused by a tap, rather than a horizontal swipe. Should we mimic this, or "enhance" them to also scroll to elements focused via tap?

The important bit is that they also don't scroll when you slide your finger over the screen to focus them. We need to retain that behavior. And I also think we shouldn't scroll if you focus the node directly by tapping it. It just seems confusing and unexpected that the screen content changes in these situations.

**Comment by nscobie** [comment]
> [Actually,] this should only happen if you land there by swiping left/right. If the user explores the screen by sliding his finger over the screen (explore by touch), the content of the screen should not change as that would be very confusing. Therefore, if the focus lands on a partially visible node by sliding your finger over it it should not scroll the list.

Ah, that makes a lot of sense in hindsight, I should have reached out to you sooner about that detail, thanks for the explanation.

Regarding the [iOS part of this](https://github.com/flutter/flutter/issues/61624) (our primary use case), it sounds like we're unfortunately blocked by platform limitations, since we're unable to distinguish between touch exploration and linear navigation when an element is focused.

Regarding the [Android part of this](https://github.com/flutter/flutter/issues/36307), it sounds like we would want the `GridView` use case to be properly implemented by TalkBack, rather than trying to patch that issue on our side.

Can you think of any other uses for having an `isPartiallyHidden` `SemanticsFlag` given the limitations noted above, or should we close this issue/proposal? I am unfamiliar with the other engine implementations for web/desktop/fuchsia, do you know if this would be helpful for accessibility on those platforms? If not, my hunch is we should close this for now and revisit it in future if need be.

**Comment by goderbauer** [comment]
> Can you think of any other uses for having an isPartiallyHidden SemanticsFlag given the limitations noted above, or should we close this issue/proposal? I am unfamiliar with the other engine implementations for web/desktop/fuchsia, do you know if this would be helpful for accessibility on those platforms? If not, my hunch is we should close this for now and revisit it in future if need be.

I would suggest closing this and revisiting it when we have an actual need for the flag.
